core/client/fs/src/main/java/alluxio/client/block/stream/NaiveSharedGrpcDataReader.java

```diff
@@ -0,0 +1,171 @@
+    mBlockId = readRequest.getBlockId();
+    mCachedDataReader = reader;
+  }
+
+  @Override
+  public long pos() {
+    return mPosToRead;
+  }
+
+  /**
+   * Seeks to a specific position.
+   *
+   * @param pos the position to seek to
+   */
+  public void seek(long pos) {
+    mPosToRead = pos;
+  }
+
+  @Override
+  public DataBuffer readChunk() throws IOException {
+    int index = (int) (mPosToRead / mChunkSize);
+    DataBuffer chunk = mCachedDataReader.readChunk(index);
+    if (chunk == null) {
+      return null;
+    }
+    ByteBuffer bb = chunk.getReadOnlyByteBuffer();
+    // Force to align to chunk size
+    bb.position((int) (mPosToRead % mChunkSize));
+    mPosToRead += mChunkSize - mPosToRead % mChunkSize;
+
+    return new NioDataBuffer(bb, bb.remaining());
+  }
+
+  @Override
+  public DataBuffer readChunkIfReady() throws IOException {
+    // I'm naive, I'm reading chunks anyway
+    return readChunk();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mCachedDataReader.deRef() == 0) {
+      try (LockResource lockResource = new LockResource(BLOCK_LOCK.writeLock())) {
+        if (mCachedDataReader.getRefCount() == 0) {
+          mCachedDataReader.close();
+          BLOCK_READERS.remove(mBlockId);
+        }
+      }
+    }
+  }
+
+  /**
+   * Factory class to create {@link NaiveSharedGrpcDataReader}s.
+   */
+  public static class Factory implements DataReader.Factory {
+    private final FileSystemContext mContext;
+    private final WorkerNetAddress mAddress;
+    private final ReadRequest mReadRequestPartial;
+    private final long mBlockSize;
+
+    /**
+     * Creates an instance of {@link NaiveSharedGrpcDataReader.Factory} for block reads.
+     *
+     * @param context the file system context
+     * @param address the worker address
+     * @param readRequestPartial the partial read request
+     * @param blockSize the block size
+     */
+    public Factory(FileSystemContext context, WorkerNetAddress address,
+        ReadRequest readRequestPartial, long blockSize) {
+      mContext = context;
+      mAddress = address;
+      mReadRequestPartial = readRequestPartial;
+      mBlockSize = blockSize;
+    }
+
+    @Override
+    public DataReader create(long offset, long len) throws IOException {
+      long blockId = mReadRequestPartial.getBlockId();
+      NaiveCachedGrpcDataReader reader;
+      try (LockResource lockResource = new LockResource(BLOCK_LOCK.writeLock())) {
+        reader = BLOCK_READERS.get(blockId);
+        if (reader == null) {
+          // I'm naive, I always read from 0 and read the whole block
+          ReadRequest cacheRequest = mReadRequestPartial
+              .toBuilder().setOffset(0).setLength(mBlockSize).build();
+          reader = new NaiveCachedGrpcDataReader
+              .Factory(mContext, mAddress, cacheRequest).create();
+          BLOCK_READERS.put(blockId, reader);
+        }
+
+        reader.ref();
+      }
```

Comment:
Changed to use an array of locks
```
try (LockResource lockResource = new LockResource(
          BLOCK_LOCKS[(int) (mBlockId % BLOCK_LOCKS.length)].writeLock())) {
```
and changed the readers hashmap to ConcurrentHashMap.
i didn't change to use `computeIfAbsent` because the `BufferCachingGrpcDataReader` creation will throw IOException which the compute function cannot process well.